### PR TITLE
Support lettuce plugin

### PR DESCRIPTION
**Pull request by zhaoyuguang**
Please answer these questions before submitting pull request

- Why submit this pull request?
- [ ] Bug fix
- [ ] New feature provided
- [x] Improve performance

___
### New feature or improvement
- Describe the details and related test reports.

### CODE REVIEW

**Comment by WithLin** [comment]
Please provide automatical test cases. Do you know where are they?

**Comment by zhaoyuguang** [comment]
Testcase URL https://github.com/SkywalkingTest/skywalking-agent-testcases/pull/48

**Comment by wu-sheng** [comment]
@ascrutae please run tests

**Comment by wu-sheng** [comment]
Tests look good. Please review. @WithLin @ascrutae too.

@zhaoyuguang Please set the supported list right

**Comment by zhaoyuguang** [comment]
Complete.

**Comment by zhaoyuguang** [comment]
I am compatible with the low version. @ascrutae please run tests~

**Comment by ascrutae** [comment]
I think that you maybe miss buried point about to catch the exception when the `io.lettuce.core.protocol. AsyncCommand#onComplete(BiConsumer<? super T,Throwable> action)` method invoked.

**Comment by wu-sheng** [comment]
@ascrutae `buried point` should be called `intercept point`. Please keep the concept consistently.

**Comment by zhaoyuguang** [comment]
> I think that you maybe miss buried point about to catch the exception when the `io.lettuce.core.protocol. AsyncCommand#onComplete(BiConsumer<? super T,Throwable> action)` method invoked.

I don not konw why intercept AsyncCommand#onComplete method, please help me~~

**Comment by wu-sheng** [comment]
I think the question looks like around, could you catch the exception if exception occurs in redis access.

**Comment by ascrutae** [comment]
Here are the reasons: 
* The javadoc of  `onComplete(BiConsumer<? super T,Throwable> action)` show that  it will send the command result or the failure resulting from command completion. so I think you'd better to catch the exception.

* The `onComplete` method is a callback method. and it maybe running in another thread. So if  someone write logic in this callback method, it maybe broke the trace.

**Comment by zhaoyuguang** [comment]
The test case added the AsyncCommand#onComplete method, tested the 5.1.3.RELEASE, 5.0.4.RELEASE, 5.0.0.RELEASE version, no trace breaks, please help run testcase at all versions again, Thanks for your hardwork! @ascrutae

**Comment by wu-sheng** [comment]
I put a comment in your test case, from my understanding, if you didn't do capture/continued, the trace could be broken.

**Comment by zhaoyuguang** [comment]
Thanks for your guidance, lettuce versions 5.0.0~5.1.3 maven compiler by java8, I will put it in the optional plugin, so I will open [WIP] again.

**Comment by zhaoyuguang** [comment]
All right!

**Comment by zhaoyuguang** [comment]
After everyone's guidance, I finished the lettuce~

**Comment by wu-sheng** [comment]
Thanks for the update. @ascrutae Please run the tests, if pass, code review could begin.

File: apm-sniffer/optional-plugins/lettuce-5.x-plugin/pom.xml [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-compiler-plugin</artifactId>
+                <configuration>
+                    <source>1.8</source>
```
Comment (zhaoyuguang) [inline apm-sniffer/optional-plugins/lettuce-5.x-plugin/pom.xml]: Is it correct here?

File: docs/en/setup/service-agent/java-agent/README.md [hunk 2]
```diff
@@ -87,6 +87,7 @@ Now, we have the following known optional plugins.
 * [Trace Oracle and Resin](agent-optional-plugins/Oracle-Resin-plugins.md)
 * [Filter traces through specified endpoint name patterns](agent-optional-plugins/trace-ignore-plugin.md)
 * Gson serialization lib in optional plugin folder
+* Lettuce 5.x(JRE1.8+) in optional plugin folder
```
Comment (zhaoyuguang) [inline docs/en/setup/service-agent/java-agent/README.md]: Am I writing the correct version of Lettuce?

**Comment by wu-sheng** [comment]
@zhaoyuguang Do you set some expected data wrong? The test results look strange to me.

File: apm-sniffer/optional-plugins/lettuce-5.x-plugin/pom.xml (hunk 1 above)
Comment (wu-sheng) [reply apm-sniffer/optional-plugins/lettuce-5.x-plugin/pom.xml]: If CI passed, then yes.

File: docs/en/setup/service-agent/java-agent/README.md (hunk 2 above)
Comment (wu-sheng) [reply docs/en/setup/service-agent/java-agent/README.md]: I think so.

File: apm-sniffer/optional-plugins/lettuce-5.x-plugin/pom.xml (hunk 1 above)
Comment (wu-sheng) [reply apm-sniffer/optional-plugins/lettuce-5.x-plugin/pom.xml]: Also, I think a simple line properties should be good too.
```
    <properties>
          <compiler.version>1.8</compiler.version>
    </properties>
```

**Comment by zhaoyuguang** [comment]
I Commented out the asynchronous call code, temporary automation cannot be overwritten.

**Comment by ascrutae** [comment]
@zhaoyuguang Could you post me the operation name of AsyncCommand? I had see the expected data and found that the operation name seem like `AsyncCommand/onComplete` when the user called `set` method, I think this operation name is not friendly.  If so, could you modify the operation name ?

FIY @wu-sheng

**Comment by wu-sheng** [comment]
@ascrutae @zhaoyuguang If you could set the callback operationName like, `AsyncSet/onComplete`, this should be better. If there is an implementation level block, please let us known.

**Comment by zhaoyuguang** [comment]
> @zhaoyuguang Could you post me the operation name of AsyncCommand? I had see the expected data and found that the operation name seem like `AsyncCommand/onComplete` when the user called `set` method, I think this operation name is not friendly. If so, could you modify the operation name ?
> 
> FIY @wu-sheng

image
operationName  seem like `AsyncCommand/onComplete` That's the truth.

**Comment by ascrutae** [comment]
I mean what  the operation name when the code invoke the `set` method of `AsyncCommand`, Not the callback operation name. the operation name of call back is fine.  IMO, the operation name seem like `Lettuce/set`, and could you post it to me?

File: apm-sniffer/optional-plugins/lettuce-5.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/lettuce/v5/AsyncCommandMethodInterceptor.java [hunk 3]
```diff
@@ -37,7 +38,8 @@
     @SuppressWarnings("unchecked")
     public void beforeMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes,
                              MethodInterceptResult result) throws Throwable {
-        AbstractSpan span = ContextManager.createLocalSpan("AsyncCommand/onComplete");
+        AsyncCommand asyncCommand = (AsyncCommand) objInst;
+        AbstractSpan span = ContextManager.createLocalSpan("AsyncCommand/" + asyncCommand.getType().name() + "/onComplete");
```
[the remark this answers is not in the document]
Comment (zhaoyuguang) [reply apm-sniffer/optional-plugins/lettuce-5.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/lettuce/v5/AsyncCommandMethodInterceptor.java]: I am modifying it again.

**Comment by zhaoyuguang** [comment]
Complete.

**Pull request merged**